source/common/http/utility.cc

```diff
@@ -781,6 +782,65 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
   encodeLocalReply(is_reset, std::move(prepared_local_reply));
 }
 
+bool Utility::remoteAddressIsTrustedProxy(
+    const Envoy::Network::Address::InstanceConstSharedPtr& remote,
+    const std::vector<Network::Address::CidrRange> trusted_cidrs) {
+  for (const auto& cidr : trusted_cidrs) {
+    if (cidr.isInRange(*remote.get())) {
+      return true;
+    }
+  }
+  return false;
+}
+
+Utility::GetLastAddressFromXffInfo Utility::getLastNonTrustedAddressFromXFF(
+    const Http::RequestHeaderMap& request_headers,
+    const std::vector<Network::Address::CidrRange> trusted_cidrs) {
+  const auto xff_header = request_headers.ForwardedFor();
+  if (xff_header == nullptr) {
+    return {nullptr, false};
+  }
+
+  absl::string_view xff_string(xff_header->value().getStringView());
+  static constexpr absl::string_view separator(",");
+
+  const auto xff_entries = StringUtil::splitToken(xff_string, separator, false, true);
+  Network::Address::InstanceConstSharedPtr last_valid_addr;
+
+  for (auto it = xff_entries.rbegin(); it != xff_entries.rend(); it++) {
+    auto addr = Network::Utility::parseInternetAddressNoThrow(std::string(*it));
+    if (addr == nullptr) {
+      continue;
+    }
+    last_valid_addr = addr;
+
+    bool trusted = false;
+    for (const auto& cidr : trusted_cidrs) {
+      if (cidr.isInRange(*addr.get())) {
+        trusted = true;
+        break;
+      }
+    }
+
+    if (trusted) {
+      continue;
+    }
+
+    // If we reach here we found a non-trusted address
+    if (Network::Utility::isInternalAddress(*addr.get()) ||
+        Network::Utility::isLoopbackAddress(*addr.get())) {
+      // If the next IP after skipping all trusted IPs is internal or loopback
+      // we cannot determine the true remote IP: Either a proxy didn't add its IP,
+      // or XFF is malformed.
+      return {nullptr, false};
+    }
+    return {addr, true};
```

Comment:
I don't think we want to just return true here or below.  I believe the final boolean here is "do we trust the whole chain such that we can say this is an internal request".   I don't _think_ we want to equate "we have a chain of trusted cidr addresses" with "we have a trusted chain of internal addresses".  I think we should only allow treating the chain as internal iff the direct connected host is internal and every hop is not only cidr-trusted but internal.

Reply:
Oof, you're right, this is definitely the wrong thing to do here. And now the more I read on what "internal" means in Envoy the more confused I'm getting 😅. The return in `getLastAddressFromXFF` is also pretty confusing.

I've read through some history in #2559 and #2590 to try and figure out what the correct answer is here. What should "internal" mean when you're saying "I trust these CIDRs"? Internal doesn't always have to mean RFC1918 or ULA, for example.

To give a concrete example for my use-case for using this feature, I want to add my CDN's ranges, and our own internal proxy IPs (RFC1918), but _not_ the whole internal network; anything else internal should be untrusted.

If Envoy's definition of "internal" is simply "RFC1918 or ULA", perhaps while looping through `trusted_cidrs`:
- Set a boolean marking if any CIDR is not internal;
- On this line I can exchange `true` with `isInternalAddress(addr)`;
- On the final line use the bool from point 1

WDYT? I'm still not convinced I've got this entirely right.

Reply:
for "internal" I usually read "sent by someone in my company", like I trust them to hit up /quitquitquit
for "trusted" here that's "trusted to set XFF header"
so unless you trust your CDN to verify "this is actually someone from my company sending traffic through the CDN" I'd say "not internal" :-)

Reply:
How do you think we should mark that with a list of CIDRs? I could split it into "internal_cidrs" and "external_cidrs", which feels a bit clunky but would do the job.

Reply:
I've dug through how the internal address config works, and how the `detect()` method is called on the xff extension, and my understanding is:

- The bool here is `allow_trusted_address_checks`, not "do we trust this address": https://github.com/envoyproxy/envoy/blob/ac42a62d6b8a2599c9a6101b6bb0bd11220c63e6/envoy/http/original_ip_detection.h#L36-L38
- The `allow_trusted_address_checks` bool determines whether to compare the address against `internal_address_config` in HCM: https://github.com/envoyproxy/envoy/blob/ac42a62d6b8a2599c9a6101b6bb0bd11220c63e6/source/common/http/conn_manager_utility.cc#L217-L228

So given that, I think returning `true` is the correct thing to do, and then HCM will compare the address returned against its internal address config to really determine if it's internal. Given that, it fits my use-case described above perfectly and there's no need to split the list or do extra checks here.

Does my reasoning seem sound?